### PR TITLE
State barebones

### PR DESCRIPTION
**Pull request by ziofil**
Initial `State`, `DM` and `Ket` implementation.

### CODE REVIEW

File: mrmustard/lab_dev/state.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+        return self.__class__(representation, modes=sorted(self.modes + other.modes))
+    
+    def __rshift__(self, other: CircuitComponent) -> Ket | CircuitComponent:
+        r"""self is contracted with other on matching modes and ket/bra sides.
+        It adds a bra if needed (e.g. ket >> channel is treated as
+        (ket.adjoint & ket) >> channel).
+        """
+        common = set(self.wires.output.modes).intersection(other.wires.input.modes)
+        # it's important to check common to avoid adding unnecessary adjoints
+        bras_incompatible = bool(other.wires[common].input.bra) != bool(self.wires[common].output.bra)
```
Comment (SamFerracin) [inline mrmustard/lab_dev/state.py:298]: if self is a Ket it doesn't have bras

File: mrmustard/lab_dev/circuit_components.py [hunk 2]
```diff
@@ -182,6 +164,32 @@ def __getitem__(self, idx: Union[Mode, Sequence[Mode]]):
         ret._wires = self._wires[idx]
         ret._parameter_set = self.parameter_set
         return ret
+    
+    def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""Contracts self and other (as the would in a circuit), but without adding
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components.py:169]: cool!

**Review by SamFerracin — COMMENTED** [review]
Looking good!

File: mrmustard/lab_dev/circuit_components.py [hunk 3]
```diff
@@ -182,6 +164,32 @@ def __getitem__(self, idx: Union[Mode, Sequence[Mode]]):
...
+        idx_z = self.wires[common].output.indices
+        idx_zconj = other.wires[common].input.indices
+        # 3) convert bargmann -> fock if needed
+        LEFT = self.representation
+        RIGHT = other.representation
+        if isinstance(self.representation, Bargmann) and isinstance(other.representation, Fock):
+            shape = [s if i in idx_z else None for i, s in enumerate(other.representation.shape)]
+            LEFT = Fock(self.fock(shape=shape), batched=False)  # assumes fock method is implemented
+            RIGHT = other.representation
+        if isinstance(self.representation, Fock) and isinstance(other.representation, Bargmann):
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components.py:182]:
```suggestion
        elif isinstance(self.representation, Fock) and isinstance(other.representation, Bargmann):
```

File: mrmustard/lab_dev/circuit_components.py [hunk 4]
```diff
@@ -182,6 +164,32 @@ def __getitem__(self, idx: Union[Mode, Sequence[Mode]]):
         ret._wires = self._wires[idx]
         ret._parameter_set = self.parameter_set
         return ret
+    
+    def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components.py:168]: This errors:

`from mrmustard.lab_dev.transformations import *`
`d0 = Dgate(1, modes=[0])`
`d0 @ d0.adjoint`

**Comment by ziofil** [comment]
Closing. Feel free to reopen if needed.

**Pull request closed without merging**